### ISSUE TITLE
Universal Login

### ISSUE THREAD

**Issue by dbinetti** [auth0-community/ember-simple-auth-auth0#143]
So I was reading about Universal login here: https://auth0.com/docs/hosted-pages/login

And as usual I'm totally confused by the docs.  Is this something compatible with this library?  Is it a different approach entirely?  Is it possible to migrate to Universal Login and, if so, how would we go about it?

**Comment by dbinetti** [comment]
OK, I'll cross-post responding to that outreach request we have.  We'll see how serious they are!  :-)

**Comment by dbinetti** [comment]
Actually, reopening for the cross-post.

**Comment by XaserAcheron** [comment]
Ye, best to leave it open -- there might be some stuff to do addon-side and I don't wish to wimp out. :P

**Comment by jerdog** [comment]
Saw your post @dbinetti  - removed it and replying in here. Is it possible to start a topic on Community so we can better address your questions and help the community going forward?

**Comment by EUGinsane** [comment]
I saw that the quickstart for emberjs on Auth0 website has been removed, is there any articles for this ? Thanks

**Comment by XaserAcheron** [comment]
The [README](https://github.com/auth0-community/ember-simple-auth-auth0) on this repo has a starter guide -- though this isn't really the issue thread for it.

**Comment by nbenz** [comment]
@XaserAcheron I think that's in reference to accepted answer [here](https://community.auth0.com/t/universal-login-with-ember-simple-auth/16098). @dbinetti moved this to a community thread, but the answer just points us at a guide that no longer exists with no further information.

I'm curious if anyone has gotten this working and/or found some helpful documentation.

**Comment by XaserAcheron** [comment]
OK, so I've finally taken the time to investigate Universal Login and the level of effort it would take for this addon to support it, and the answer is, thankfully, "not much". :D

The lion's share of the work (URL hash parsing) is already done, so there's mainly two things to do here:
- [ ] Add an `auth0-universal` authenticator that calls the correct auth0.js method
- [ ] Add a build flag to `ember-cli-build.js` that instructs the addon to exclude Lock from the build tree.

That second bullet point is pretty significant for Ember, since that means you no longer have to pull in all of React (eep!) to build Lock. This should also eliminate quite a few pain points surrounding the build process, and Auth0 is recommending Universal Login over embedded Lock these days anyway.

**Comment by dbinetti** [comment]
Thanks, @XaserAcheron  -- for those of us sticking with Auth0 and Ember this would be a great addition....

**Comment by XaserAcheron** [comment]
No problem! Well, hopefully I don't get distracted this time. :P

Current status: got a basic implementation of the authenticator, and now I'm working on restoring the attempted transition after redirecting back to the app, so if you try and hit `https://dummy.app/some/deep/nested/url` while unauthenticated, you end up there once Auth0 redirects you back to the app.

**Comment by XaserAcheron** [comment]
Well, I made good progress until it came time to update the unit tests -- I had to modify a function to return a promise, and when I attempt to do the same in the Lock stub for the tests, they hang forever for unknowable reasons.

I'm not sure when I'm going to be able to resolve this (no pun intended) -- the entire test suite for this project is indecipherable nonsense, as far as I'm concerned.

**Comment by XaserAcheron** [comment]
OK, I think I see what's going on -- the tests are implicitly making the assumption that the Lock instance is immediately, synchronously available as soon as `showLock` is called, which is no longer the case -- it was never a safe assumption anyway, since `showLock` has always returned a promise.

tl;dr: the tests in question are no longer valid, and I'm not sure what to do to refactor them at present.

**Comment by XaserAcheron** [comment]
Bah, never mind -- got it figured out.

Apologies for the ranting -- this part makes me grumpy. :P

**Comment by dbinetti** [comment]
Just in case your life wasn't complicated enough.... :-)

https://auth0.com/docs/universal-login/new

**Comment by XaserAcheron** [comment]
There's always something, isn't there? :P

Anyhow, I'm hoping this doesn't change the behavior any on the client (i.e. non-Auth0) side, but I'd better test with a New workflow just to confirm.

**Comment by dbinetti** [comment]
```
I'm hoping this doesn't change the behavior any on the client
```
famous last words...  :-).  Thanks for all your efforts here.

**Comment by XaserAcheron** [comment]
Thanks! But maybe thank me when it works. ;P

Speaking of, got this merged in now. I'll probably cut an alpha soonish, but I've got one other thing I want to merge in first so bear with me. Grizzly, perferably.

**Comment by XaserAcheron** [comment]
Hey @dbinetti , I just published a new beta, `v5.0.0-beta.0` (yeah, it starts at zero... maybe that's on me for not using ember-cli-release correctly :P ) -- it's got an `auth0-universal` authenticator with some [instructions](https://github.com/auth0-community/ember-simple-auth-auth0#universal-login) in the main README.md on how to use it. Hopefully it's all well, but a beta's a beta. :P

Something to note is that this is a major version candidate, since I had to change some of the API surface of the `auth0` service to support this (namely, the `showLock` methods now work asynchronously and return a promise... not that anyone ought to be calling this directly, but you never know).

The v5 series contains a breaking change (the `jwt` authorizer is gone), but there's some replacement methods and a [migration guide](https://github.com/auth0-community/ember-simple-auth-auth0#migrating-from-ember-simple-auth-auth0-v4x-to-v5x) so hopefully it's not too painful.

Phew! Took me long enough to get to this. :P

**Comment by dbinetti** [comment]
Thanks so much!!  Hoping to get to this by early next week...

**Comment by dbinetti** [comment]
So @XaserAcheron I got the chance to try this out.  Two questions:

1) Should I even be reporting these to you?  I saw you've got a new job (congrats!) and am moving on.  Not sure if you're going to be supporting this particular issue or are done.

2) If the answer to 1) is "Yes", then I'm getting a:
```
{"statusCode":403,"description":"Invalid state","name":"AnomalyDetected","code":"access_denied"}
```
error when trying to use it.  If the answer is "No", to whom are we directing issues?

Thanks!
